### PR TITLE
Add coinbase prime

### PR DESCRIPTION
**Pull request by yabirgb**
Still pending:

- Add location in db upgrade
- Add logic to process other events as history events

### CODE REVIEW

File: rotkehlchen/tests/fixtures/rotkehlchen.py [hunk 1]
```diff
@@ -679,14 +684,14 @@ def rotkehlchen_instance(
 
 @pytest.fixture
 def rotkehlchen_api_server_with_exchanges(
-        rotkehlchen_api_server,
-        added_exchanges,
-        gemini_test_base_uri,
-        gemini_sandbox_api_secret,
-        gemini_sandbox_api_key,
```
Comment (yabirgb) [inline rotkehlchen/tests/fixtures/rotkehlchen.py:686]: Typed this function because I had issues with this fixture until I figured what was wrong

File: rotkehlchen/exchanges/coinbaseprime.py [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+                try:
+                    if (event := decoding_logic(raw_event)) is None:
+                        log.warning(
+                            f'Wont process event {raw_event} from coinbase prime. Skipping',
+                        )
+                    else:
+                        result.append(event)
+                except DeserializationError as e:
+                    self.msg_aggregator.add_error(
+                        f'Failed to process coinbase prime trade due to {e}. Skipping entry...',
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbaseprime.py:278]: Is it always trades here?

```suggestion
                        f'Failed to process coinbase prime event due to {e}. Skipping entry...',
```

File: rotkehlchen/exchanges/coinbaseprime.py [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+
+            if (
+                response['pagination']['has_next'] is True and
+                (new_cursor := response['pagination'].get('next_cursor', '')) != ''
+            ):
+                query_params['cursor'] = new_cursor
+            else:
+                break
+
+        return result  # type: ignore  # mypy doesn't detect that the return type here is defined by the function used
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbaseprime.py:289]: It can't since `self._api_query` returns a dict, a key of whose is different. You are not showing this difference with an overload at that method.

File: rotkehlchen/exchanges/coinbaseprime.py [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+    def _get_portfolio_ids(self) -> list[str]:
+        """
+        Get id of the different portfolios linked to the api keys.
+        May raise: RemoteError
+        """
+        data: dict[str, list] = self._api_query(module='portfolios')
+        try:
+            ids = [portfolio['id'] for portfolio in data.get('portfolios', [])]
+        except KeyError:
+            log.error('Malformed porfolios ids response in coinbase prime. Missing key id')
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbaseprime.py:300]: 1. Missing info from the log
2. If we get here, it's a bug as `ids` is not defined. What do you want to do? Return empty list or raise?

```suggestion
            log.error(f'Malformed porfolios ids response in coinbase prime. Missing key id: {data["portfolios"]}')
            return or raise           
```

File: rotkehlchen/exchanges/coinbaseprime.py [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+                )
+            except RemoteError as e:
+                log.error(f'Failed to query CoinbasePrime balances due to {e}')
+                return None, 'Request to CoinbasePrime failed to fetch balances'
+
+            for balance_entry in balances_query['balances']:
+                try:
+                    total_balance = ZERO
+                    for balance_key in (
+                        'amount',  # includes holds fields as per the docs
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbaseprime.py:333]: holds fields? That sounds like broken English.

File: rotkehlchen/exchanges/coinbaseprime.py [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+                    # ignore empty balances. Coinbase returns zero balances for everything
+                    # a user does not own
+                    if total_balance == ZERO:
+                        continue
+
+                    asset = asset_from_coinbase(balance_entry['symbol'])
+                    try:
+                        usd_price = Inquirer.find_usd_price(asset=asset)
+                    except RemoteError as e:
+                        self.msg_aggregator.add_error(
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbaseprime.py:353]: Do we really need to add an aggregator error in those? It's just an oracle error which can happen. Shuldn't we just log and continue?

File: rotkehlchen/exchanges/kucoin.py [hunk 7]
```diff
@@ -165,15 +165,12 @@ def __init__(
...
-    def update_passphrase(self, new_passphrase: str) -> None:
-        self.api_passphrase = new_passphrase
-
     def edit_exchange_credentials(self, credentials: ExchangeAuthCredentials) -> bool:
         changed = super().edit_exchange_credentials(credentials)
         if credentials.api_key is not None:
             self.session.headers.update({'KC-API-KEY': self.api_key})
         if credentials.passphrase is not None:
-            self.update_passphrase(credentials.passphrase)
+            self.api_passphrase = credentials.passphrase
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/kucoin.py:173]: why this change?

File: rotkehlchen/tests/fixtures/exchanges/coinbase.py [hunk 8]
```diff
@@ -9,8 +12,20 @@ def function_scope_coinbase(
...
-    mock = create_test_coinbase(
+    return create_test_coinbase(
         database=database,
         msg_aggregator=function_scope_messages_aggregator,
     )
-    return mock
+
+
+@pytest.fixture
+def function_scope_coinbase_prime(
```
Comment (LefterisJP) [inline rotkehlchen/tests/fixtures/exchanges/coinbase.py:22]:
```suggestion
def function_scope_coinbaseprime(
```

File: rotkehlchen/exchanges/coinbaseprime.py [hunk 9]
```diff
@@ -0,0 +1,458 @@
...
+
+    def sign(self, timestamp: Timestamp, url_path: str) -> bytes:
+        """Sign requests to coinbase prime"""
+        message = f'{timestamp}GET{url_path}'
+        hmac_message = hmac.digest(self.secret, message.encode(), hashlib.sha256)
+        return base64.b64encode(hmac_message)
+
+    def _api_query(
+            self,
+            module: Literal['portfolios'],
```
[the remark this answers is not in the document]
Comment (yabirgb) [reply rotkehlchen/exchanges/coinbaseprime.py:216]: they have `portfolios` inside the account so whatever you query (for what we need) is always relative to a porfolio and the paths are relative to the potfolio endpoint.

File: rotkehlchen/assets/converters.py [hunk 10]
```diff
@@ -207,7 +207,7 @@ def asset_from_coinbase(cb_name: str, time: Timestamp | None = None) -> AssetWit
 
     return symbol_to_asset_or_token(GlobalDBHandler.get_assetid_from_exchange_name(
         exchange=Location.COINBASE,
-        symbol=cb_name,
+        symbol=cb_name.upper(),
```
[the remark this answers is not in the document]
Comment (yabirgb) [reply rotkehlchen/assets/converters.py:210]: Coinbase returns already capital strings and so does coinbase prime except for the balance endpoint. This doesn't affect to coinbase. Checked it with

```
python pytestgeventwrapper.py rotkehlchen/tests/exchanges/test_coinbase.py::test_coverage_of_products
```

File: rotkehlchen/exchanges/coinbaseprime.py (hunk 3 above)
Comment (yabirgb) [reply rotkehlchen/exchanges/coinbaseprime.py:306]: The return of `_api_query` doesn't affect to the variable result. Notice that I'm using `event := decoding_logic(raw_event)` and then `result.append(event)` is the signature of `decoding_logic` what should affect the signature of result

File: rotkehlchen/exchanges/coinbaseprime.py [hunk 11]
```diff
@@ -0,0 +1,458 @@
...
+            'Content-Type': 'application/json',
+            'X-CB-ACCESS-KEY': self.api_key,
+            'X-CB-ACCESS-PASSPHRASE': self.api_passphrase,
+        })
+
+    def validate_api_key(self) -> tuple[bool, str]:
+        try:
+            self._get_portfolio_ids()
+        except RemoteError as e:
+            return False, str(e)
```
[the remark this answers is not in the document]
Comment (yabirgb) [reply rotkehlchen/exchanges/coinbaseprime.py:190]: The str(e) is propagated and logged/shown to the user

File: rotkehlchen/assets/converters.py (hunk 10 above)
Comment (LefterisJP) [inline rotkehlchen/assets/converters.py:210]: Please add a comment explaining what you said here. You can't put an `uppercase()` without explaining why. In the future nobody will remember why this is here.

File: rotkehlchen/exchanges/coinbaseprime.py [hunk 12]
```diff
@@ -0,0 +1,477 @@
...
+def _process_trade(trade_data: dict[str, Any]) -> Trade | None:
+    """Process trade from coinbase prime. Returns None if the order can't be processed
+    May raise:
+    - DeserializationError
+    """
+    try:
+        if trade_data['status'] != 'FILLED':
+            return None
+
+        if type(trade_data['product_id']) is not str and '-' not in trade_data['product_id']:
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbaseprime.py:66]: Don't check type with `type`. Always `isinstance`. Various benefits such as also subclasses etc. Google or ask claude for details

```suggestion
        if not isinstance(trade_data['product_id'], str) and '-' not in trade_data['product_id']:
```

File: rotkehlchen/exchanges/coinbaseprime.py [hunk 13]
```diff
@@ -0,0 +1,477 @@
...
+            params: dict[str, Any] | None = None,
+    ) -> dict[str, Any]:
+        uri = f'{PRIME_BASE_URL}/{module}'
+        if path != '':
+            uri += f'/{path}'
+
+        url_path = urlparse(uri).path
+        timestamp = ts_now()
+        self.session.headers.update({
+            'X-CB-ACCESS-TIMESTAMP': str(timestamp),
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbaseprime.py:227]: nitpick

```suggestion
        self.session.headers.update({
            'X-CB-ACCESS-TIMESTAMP': str((timestamp := ts_now()),
```

File: rotkehlchen/exchanges/coinbaseprime.py [hunk 14]
```diff
@@ -0,0 +1,477 @@
...
+
+            if (
+                response['pagination']['has_next'] is True and
+                (new_cursor := response['pagination'].get('next_cursor', '')) != ''
+            ):
+                query_params['cursor'] = new_cursor
+            else:
+                break
+
+        return result  # type: ignore  # mypy doesn't detect that the return type here is defined by the function used
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbaseprime.py:306]: For this don't you think there is a way to do it? Don't want you to spend time here, but some times it's missing some obvious typing thing. Have you tried going in with the `reveal_type()` in various places and seeing what mypy thinks it is?

File: rotkehlchen/exchanges/coinbaseprime.py [hunk 15]
```diff
@@ -0,0 +1,477 @@
...
+        """
+        Get id of the different portfolios linked to the api keys.
+        May raise: RemoteError
+        """
+        data: dict[str, list] = self._api_query(module='portfolios')
+        try:
+            ids = [portfolio['id'] for portfolio in data.get('portfolios', [])]
+        except KeyError as e:
+            raise RemoteError(
+                f'Malformed porfolios ids response in coinbase prime. Missing key {e} in {data}',
```
Comment (LefterisJP) [inline rotkehlchen/exchanges/coinbaseprime.py:318]: This is too much for a remote error. Remote Errors should not contain {data}. Log it instead. As far as I remember `RemoteError` text gets propagated to the user.

File: rotkehlchen/exchanges/coinbaseprime.py (hunk 14 above)
Comment (yabirgb) [reply rotkehlchen/exchanges/coinbaseprime.py:305]: https://github.com/orgs/rotki/projects/11?pane=issue&itemId=85419257 created a note

**Review by LefterisJP — APPROVED** [review]
LGTM
